lib/ProductOpener/TaxonomySuggestions.pm

```diff
@@ -333,45 +377,45 @@ sub filter_suggestions_matching_string ($tags_ref, $tagtype, $search_lc, $string
 			# just_synonyms are not real entries
 			next if defined $just_synonyms{$tagtype}{$canon_tagid};
 
-			my $tag;    # this is the content string
-			my $tagid;    # this is the tag
-
-			# search if the tag exists in target language
-			if (defined $translations_to{$tagtype}{$canon_tagid}{$search_lc}) {
-
-				$tag = $translations_to{$tagtype}{$canon_tagid}{$search_lc};
-				# TODO: explain why $tagid can be different from $canon_tagid
-				$tagid = get_string_id_for_lang($search_lc, $tag);
-
-				# add language prefix if we are not searching current interface language
-				if (not($search_lc eq $original_lc)) {
-					$tag = $search_lc . ":" . $tag;
+			# We will match synonyms in the search language, and in the wildcard xx: language
+			my $tag = display_taxonomy_tag($search_lc, $tagtype, $canon_tagid);
+			my $tag_xx = display_taxonomy_tag("xx", $tagtype, $canon_tagid);
+
+			# Build a list of normalized synonyms in the search language and the wildcard xx: language
+			my @synonyms_ids = map {get_string_id_for_lang($search_lc, $_)} (
+				@{deep_get(\%synonyms_for, $tagtype, $search_lc, get_string_id_for_lang($search_lc, $tag)) || []},
+				@{deep_get(\%synonyms_for, $tagtype, "xx", get_string_id_for_lang("xx", $tag_xx)) || []}
+			);
+
+			# check how well the synonyms match the input string
+			my $best_match = best_match($stringid, $fuzzystringid, \@synonyms_ids);
+
+			$log->debug(
+				"synonyms_ids for canon_tagid",
+				{
+					tagtype => $tagtype,
+					canon_tagid => $canon_tagid,
+					tag => $tag,
+					synonym_ids => \@synonyms_ids,
+					best_match => $best_match
 				}
-			}
-			# also search for special language code "xx" which is universal
-			elsif (defined $translations_to{$tagtype}{$canon_tagid}{xx}) {
-				$tag = $translations_to{$tagtype}{$canon_tagid}{xx};
-				$tagid = get_string_id_for_lang("xx", $tag);
-			}
+			) if $log->is_debug();
 
-			if (defined $tag) {
-				# matching at start, best matches
-				if ($tagid =~ /^$stringid/) {
-					push @suggestions, $tag;
-					# only matches at start are considered
-					$suggestions_count++;
-				}
-				# matching inside
-				elsif ($tagid =~ /$stringid/) {
-					push @suggestions_c, $tag;
-				}
-				# fuzzy match
-				elsif ($tagid =~ /$fuzzystringid/) {
-					push @suggestions_f, $tag;
-				}
-				# end as soon as we got enough
+			# matching at start, best matches
+			if ($best_match eq "start") {
+				push @suggestions, $tag;
+				# count matches at start so that we can return only if we have enough matches
+				$suggestions_count++;
 				last if $suggestions_count >= $limit;
 			}
+			# matching inside
+			elsif ($best_match eq "inside") {
+				push @suggestions_c, $tag;
+			}
+			# fuzzy match
+			elsif ($best_match eq "fuzzy") {
+				push @suggestions_f, $tag;
+			}
 		}
 	}
```

Comment:
Why not. In the worse case scenario (huge taxonomy like categories, and a text query that does not match anything so we have to go through everything, it takes about 0.5 seconds in prod). I'm not sure we will gain much on smaller taxonomies, but we only have a bit of space to lose.

Another option is to use memcached (so that the cache is shared between processes), but then it might be tricky when we update things and get old cache results even after restarting Apache.

Reply:
Oh yes, I didn't remember we where using multiple processes (and quite a lot) so memcached would really be needed. Only thing is that I didn't see ready to use memoize using memcached with a RLU strategy (MRU would be even better).

Reply:
Right, I chose to use memcached, I'm refactoring some stuff around it.